src/wallet/scriptpubkeyman.h

```diff
@@ -0,0 +1,479 @@
+        int nVersion = s.GetVersion();
+        if (!(s.GetType() & SER_GETHASH))
+            READWRITE(nVersion);
+        READWRITE(nTime);
+        READWRITE(vchPubKey);
+        if (ser_action.ForRead()) {
+            try {
+                READWRITE(fInternal);
+            }
+            catch (std::ios_base::failure&) {
+                /* flag as external address if we can't read the internal boolean
+                   (this will be the case for any wallet before the HD chain split version) */
+                fInternal = false;
+            }
+            try {
+                READWRITE(m_pre_split);
+            }
+            catch (std::ios_base::failure&) {
+                /* flag as postsplit address if we can't read the m_pre_split boolean
+                   (this will be the case for any wallet that upgrades to HD chain split)*/
+                m_pre_split = false;
+            }
+        }
+        else {
+            READWRITE(fInternal);
+            READWRITE(m_pre_split);
+        }
+    }
+};
+
+/*
+ * A class implementing ScriptPubKeyMan manages some (or all) scriptPubKeys used in a wallet.
+ * It contains the scripts and keys related to the scriptPubKeys it manages.
+ * A ScriptPubKeyMan will be able to give out scriptPubKeys to be used, as well as marking
+ * when a scriptPubKey has been used. It also handles when and how to store a scriptPubKey
+ * and it's related scripts and keys, including encryption.
+ */
+class ScriptPubKeyMan
+{
+private:
+    WalletStorage& storage;
+
+protected:
+    std::shared_ptr<WalletDatabase> m_database;
+
+public:
+    ScriptPubKeyMan(WalletStorage& storage) : storage(storage), m_database(storage.GetDatabase()) {}
+
+    const std::string GetDisplayName() const { return storage.GetDisplayName(); }
+    bool IsWalletFlagSet(uint64_t flag) const { return storage.IsWalletFlagSet(flag); }
+    void SetWalletFlag(uint64_t flag) { return storage.SetWalletFlag(flag); }
+    void UnsetWalletFlagWithDB(WalletBatch& batch, uint64_t flag) { return storage.UnsetWalletFlagWithDB(batch, flag); }
+    bool CanSupportFeature(enum WalletFeature wf) const { return storage.CanSupportFeature(wf); }
+    void SetMinVersion(enum WalletFeature wf) { return storage.SetMinVersion(wf); }
+    const CKeyingMaterial& GetEncryptionKey() const { return storage.GetEncryptionKey(); }
+    bool IsLocked() const { return storage.IsLocked(); }
+
+    virtual ~ScriptPubKeyMan() {};
+    virtual bool GetNewDestination(const OutputType type, CTxDestination& dest, std::string& error) { return false; }
+    virtual isminetype IsMine(const CScript& script) const { return ISMINE_NO; }
+
+    virtual bool HasEncryptedData() const { return false; }
+
+    virtual bool CheckDecryptionKey(const CKeyingMaterial& vMasterKeyIn, bool accept_no_keys = false) { return false; }
+    virtual bool Encrypt(CKeyingMaterial& vMasterKeyIn, WalletBatch* batch) { return false; }
```

Comment:
Master key argument should be a const reference, not mutable. Also these arguments could be named master_key, since there's no longer any vMasterKeyIn/vMasterKey conflict.

Reply:
Done